File: src/types/library.ts
// src/types/library.ts

export interface StarredItem {
  id: string;
  type: string; // "mod" | "modpack" | "server"
  source: string; // "modrinth" | "curseforge" | "custom"
  projectId?: string;
  title?: string;
  author?: string;
  snapshot: string; // JSON string payload
  state: string;    // JSON string payload
  meta: string;     // JSON string payload
  createdAt: number;
  updatedAt: number;
}

export interface SnapshotPayload {
  title: string;
  iconUrl?: string;
  author?: string;
  description?: string;
  loaders?: string[];
  categories?: string[];
  version?: string;
  updatedAt?: string;
}

export interface StatePayload {
  installedVersion?: string;
  lastKnownVersion?: string;
  lastKnownUpdatedAt?: string;
  hasUpdate: boolean;
  lastCheckedAt?: number;
}

export interface MetaPayload {
  createdAt: number;
  updatedAt: number;
  tags?: string[];
  note?: string;
  pinned?: boolean;
  archived?: boolean;
}

export interface Collection {
  id: string;
  name: string;
  description?: string;
  type: string; // "group" | "modpack" | "favorite"
  coverImage?: string;
  sortOrder: number;
  createdAt: number;
  updatedAt: number;
}

export interface CollectionItem {
  id: string;
  collectionId: string;
  itemId: string;
  position: number;
  extra?: string; // JSON string payload
  createdAt: number;
}

export interface CollectionItemExtra {
  lockedVersion?: string;
  required?: boolean;
}

File: src-tauri/src/domain/library.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct StarredItem {
    pub id: String,
    pub r#type: String, // "mod" | "modpack" | "server"
    pub source: String, // "modrinth" | "curseforge" | "custom"
    pub project_id: Option<String>,
    pub title: Option<String>,
    pub author: Option<String>,
    pub snapshot: String, // JSON payload representing immutable core data
    pub state: String,    // JSON payload representing volatile status like version, hasUpdate
    pub meta: String,     // JSON payload representing user specific tags, notes
    pub created_at: i64,
    pub updated_at: i64,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct Collection {
    pub id: String,
    pub name: String,
    pub description: Option<String>,
    pub r#type: String, // "group" | "modpack" | "favorite"
    pub cover_image: Option<String>,
    pub sort_order: i32,
    pub created_at: i64,
    pub updated_at: i64,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct CollectionItem {
    pub id: String,
    pub collection_id: String,
    pub item_id: String,
    pub position: i32,
    pub extra: Option<String>, // JSON specific to integration logic
    pub created_at: i64,
}
